test/001_highsnr_stdio_audio/test_tx.py

```diff
@@ -0,0 +1,163 @@
+
+
+
+# open codec2 instance        
+freedv = cast(codec2.api.freedv_open(MODE), c_void_p)
+
+# get number of bytes per frame for mode
+bytes_per_frame = int(codec2.api.freedv_get_bits_per_modem_frame(freedv)/8)
+payload_bytes_per_frame = bytes_per_frame -2
+
+# init buffer for data
+n_tx_modem_samples = codec2.api.freedv_get_n_tx_modem_samples(freedv)
+mod_out = create_string_buffer(n_tx_modem_samples * 2)
+
+# init buffer for preample
+n_tx_preamble_modem_samples = codec2.api.freedv_get_n_tx_preamble_modem_samples(freedv)
+mod_out_preamble = create_string_buffer(n_tx_preamble_modem_samples * 2)
+
+# init buffer for postamble
+n_tx_postamble_modem_samples = codec2.api.freedv_get_n_tx_postamble_modem_samples(freedv)
+mod_out_postamble = create_string_buffer(n_tx_postamble_modem_samples * 2)
+
+
+# create buffer for data
+buffer = bytearray(payload_bytes_per_frame) # use this if CRC16 checksum is required ( DATA1-3)
+buffer[:len(data_out)] = data_out # set buffersize to length of data which will be send
+
+# create crc for data frame - we are using the crc function shipped with codec2 to avoid 
+# crc algorithm incompatibilities
+crc = ctypes.c_ushort(codec2.api.freedv_gen_crc16(bytes(buffer), payload_bytes_per_frame))     # generate CRC16
+crc = crc.value.to_bytes(2, byteorder='big') # convert crc to 2 byte hex string
+buffer += crc        # append crc16 to buffer    
+
+print(f"TOTAL BURSTS: {N_BURSTS} TOTAL FRAMES_PER_BURST: {N_FRAMES_PER_BURST}", file=sys.stderr)
+
+for i in range(1,N_BURSTS+1):
+
+    # write preamble to txbuffer
+    codec2.api.freedv_rawdatapreambletx(freedv, mod_out_preamble)
+    txbuffer = bytes(mod_out_preamble)
+
+    # create modulaton for N = FRAMESPERBURST and append it to txbuffer
+    for n in range(1,N_FRAMES_PER_BURST+1):
+
+        data = (ctypes.c_ubyte * bytes_per_frame).from_buffer_copy(buffer)
+        codec2.api.freedv_rawdatatx(freedv,mod_out,data) # modulate DATA and save it into mod_out pointer 
+
+        txbuffer += bytes(mod_out)
+        
+        print(f"TX BURST: {i}/{N_BURSTS} FRAME: {n}/{N_FRAMES_PER_BURST}", file=sys.stderr)
+    
+    # append postamble to txbuffer          
+    codec2.api.freedv_rawdatapostambletx(freedv, mod_out_postamble)
+    txbuffer += bytes(mod_out_postamble)
+
+    # append a delay between bursts as audio silence
+    samples_delay = int(MODEM_SAMPLE_RATE*DELAY_BETWEEN_BURSTS)
+    mod_out_silence = create_string_buffer(samples_delay*2)
+    txbuffer += bytes(mod_out_silence)
+    #print(f"samples_delay: {samples_delay} DELAY_BETWEEN_BURSTS: {DELAY_BETWEEN_BURSTS}", file=sys.stderr)
+
+    # resample up to 48k (resampler works on np.int16)
+    x = np.frombuffer(txbuffer, dtype=np.int16)
+    txbuffer_48k = resampler.resample8_to_48(x)
+    
+    # check if we want to use an audio device or stdout
+    if AUDIO_OUTPUT_DEVICE != -1:
+        # Gotcha: we have to convert from np.int16 to Python "bytes"
+        stream_tx.write(txbuffer_48k.tobytes())
```

Comment:
This caught me for a while - `write()` doesn't actually choke when you pass it a `np.int16` array, but it's corrupted on playback.  I tracked this one down by playing a test sinewave using `test_pa.py`

All this moving back and forth between `bytes` and `np.int16` is a candidate for refactoring.  I'm not sure if it's too messy.  It's just I understand numpy a bit better than these `byte` thingies.